File: app/src/main/java/rs/ove/crypt/proto/Mst5MediaClient.java
package rs.ove.crypt.proto;

import android.os.ParcelFileDescriptor;

/** File-descriptor facade for Rust mst5-client media streams. */
public final class Mst5MediaClient {
	public interface Observer {
		boolean isCancelled();
		void onProgress(long completed, long total);
	}

	private Mst5MediaClient() {}

	public static final class Upload {
		public final String endpoint;
		public final String publicKey;
		public final String ticket;
		public final String fileId;
		public final long size;
		public final ParcelFileDescriptor source;

		public Upload(String endpoint, String publicKey, String ticket, String fileId,
		              long size, ParcelFileDescriptor source) {
			this.endpoint = endpoint;
			this.publicKey = publicKey;
			this.ticket = ticket;
			this.fileId = fileId;
			this.size = size;
			this.source = source;
		}
	}

	public static void uploadDescriptor(String endpoint, String publicKeyB64, String ticket,
	                                    String fileId, long size, ParcelFileDescriptor source,
	                                    final Observer observer) throws Exception {
		NativeMst5.upload(endpoint, publicKeyB64, ticket, fileId, size, source, adapt(observer));
	}

	public static void uploadE2EDescriptor(String endpoint, String publicKeyB64, String ticket,
	                                      String fileId, long plaintextSize, ParcelFileDescriptor source,
	                                      long identityHandle, byte[] peer, String from, String to,
	                                      final Observer observer) throws Exception {
		NativeMst5.uploadE2E(endpoint, publicKeyB64, ticket, fileId, plaintextSize, source,
				identityHandle, peer, from, to, adapt(observer));
	}

	public static void uploadDescriptors(java.util.List<Upload> uploads, final Observer observer) throws Exception {
		if (uploads == null || uploads.isEmpty()) return;
		int count = uploads.size();
		String[] endpoints = new String[count];
		String[] publicKeys = new String[count];
		String[] tickets = new String[count];
		String[] fileIds = new String[count];
		long[] sizes = new long[count];
		int[] fds = new int[count];
		for (int i = 0; i < count; i++) {
			Upload upload = uploads.get(i);
			if (upload == null || upload.source == null) throw new java.io.IOException("media descriptor is unavailable");
			endpoints[i] = upload.endpoint;
			publicKeys[i] = upload.publicKey;
			tickets[i] = upload.ticket;
			fileIds[i] = upload.fileId;
			sizes[i] = upload.size;
			fds[i] = upload.source.getFd();
		}
		NativeMst5.Observer nativeObserver = adapt(observer);
		if (nativeObserver == null) {
			nativeObserver = new NativeMst5.Observer() {
				@Override public boolean isCancelled() { return false; }
				@Override public void onProgress(long completed, long total) {}
			};
		}
		NativeMst5.uploadBatch(endpoints, publicKeys, tickets, fileIds, sizes, fds, nativeObserver);
	}

	public static long downloadDescriptor(String endpoint, String publicKeyB64, String ticket,
	                                     String fileId, long expectedSize, ParcelFileDescriptor target,
	                                     final Observer observer) throws Exception {
		return NativeMst5.download(endpoint, publicKeyB64, ticket, fileId, expectedSize, target, adapt(observer));
	}

	public static long downloadE2EDescriptor(String endpoint, String publicKeyB64, String ticket,
	                                        String fileId, long encryptedSize, ParcelFileDescriptor target,
	                                        long identityHandle, byte[] peer, String from, String to,
	                                        final Observer observer) throws Exception {
		return NativeMst5.downloadE2E(endpoint, publicKeyB64, ticket, fileId, encryptedSize, target,
				identityHandle, peer, from, to, adapt(observer));
	}

	public static byte[] downloadE2EBytes(String endpoint, String publicKeyB64, String ticket,
	                                    String fileId, long encryptedSize, int maxBytes, long identityHandle,
	                                    byte[] peer, String from, String to) throws Exception {
		return NativeMst5.downloadE2EBytes(endpoint, publicKeyB64, ticket, fileId, encryptedSize,
				maxBytes, identityHandle, peer, from, to);
	}

	public static byte[] downloadBytes(String endpoint, String publicKeyB64, String ticket,
	                                   String fileId, long expectedSize, int maxBytes) throws Exception {
		return NativeMst5.downloadBytes(endpoint, publicKeyB64, ticket, fileId, expectedSize, maxBytes);
	}

	private static NativeMst5.Observer adapt(final Observer observer) {
		if (observer == null) return null;
		return new NativeMst5.Observer() {
			@Override public boolean isCancelled() { return observer.isCancelled(); }
			@Override public void onProgress(long completed, long total) {
				observer.onProgress(completed, total);
			}
		};
	}
}

File: app/src/main/kotlin/rs/ove/crypt/proto/Mst5MediaClient.kt
package rs.ove.crypt.proto

import android.os.ParcelFileDescriptor
import java.io.IOException

/** File-descriptor facade for Rust mst5-client media streams. */
object Mst5MediaClient {
    interface Observer { fun isCancelled(): Boolean; fun onProgress(completed: Long, total: Long) }
    class Upload(@JvmField val endpoint: String?, @JvmField val publicKey: String?, @JvmField val ticket: String?, @JvmField val fileId: String?, @JvmField val size: Long, @JvmField val source: ParcelFileDescriptor?)
    @JvmStatic @Throws(Exception::class) fun uploadE2EDescriptor(endpoint: String?, publicKeyB64: String?, ticket: String?, fileId: String?, plaintextSize: Long, source: ParcelFileDescriptor?, identityHandle: Long, peer: ByteArray?, from: String?, to: String?, observer: Observer?) = NativeMst5.uploadE2E(endpoint, publicKeyB64, ticket, fileId, plaintextSize, source, identityHandle, peer, from, to, adapt(observer))
    @JvmStatic @Throws(Exception::class) fun uploadDescriptors(uploads: List<Upload>?, observer: Observer?) {
        if (uploads.isNullOrEmpty()) return
        var total = 0L
        uploads.forEach { upload ->
            if (upload.source == null || upload.size < 0) throw IOException("media descriptor is unavailable")
            if (Long.MAX_VALUE - total < upload.size) throw IOException("media batch size overflow")
            total += upload.size
        }
        var completed = 0L
        uploads.forEach { upload ->
            val base = completed
            NativeMst5.upload(upload.endpoint, upload.publicKey, upload.ticket, upload.fileId, upload.size, upload.source,
                object : NativeMst5.Observer {
                    override fun isCancelled(): Boolean = observer?.isCancelled() == true
                    override fun onProgress(done: Long, ignored: Long) { observer?.onProgress(minOf(total, base + maxOf(0, done)), total) }
                })
            completed += upload.size
        }
    }
    @JvmStatic @Throws(Exception::class) fun downloadDescriptor(endpoint: String?, publicKeyB64: String?, ticket: String?, fileId: String?, expectedSize: Long, target: ParcelFileDescriptor?, observer: Observer?): Long = NativeMst5.download(endpoint, publicKeyB64, ticket, fileId, expectedSize, target, adapt(observer))
    @JvmStatic @Throws(Exception::class) fun downloadE2EDescriptor(endpoint: String?, publicKeyB64: String?, ticket: String?, fileId: String?, encryptedSize: Long, target: ParcelFileDescriptor?, identityHandle: Long, peer: ByteArray?, from: String?, to: String?, observer: Observer?): Long = NativeMst5.downloadE2E(endpoint, publicKeyB64, ticket, fileId, encryptedSize, target, identityHandle, peer, from, to, adapt(observer))
    @JvmStatic @Throws(Exception::class) fun downloadE2EBytes(endpoint: String?, publicKeyB64: String?, ticket: String?, fileId: String?, encryptedSize: Long, maxBytes: Int, identityHandle: Long, peer: ByteArray?, from: String?, to: String?): ByteArray = NativeMst5.downloadE2EBytes(endpoint, publicKeyB64, ticket, fileId, encryptedSize, maxBytes, identityHandle, peer, from, to)
    @JvmStatic @Throws(Exception::class) fun downloadBytes(endpoint: String?, publicKeyB64: String?, ticket: String?, fileId: String?, expectedSize: Long, maxBytes: Int): ByteArray = NativeMst5.downloadBytes(endpoint, publicKeyB64, ticket, fileId, expectedSize, maxBytes)
    private fun adapt(observer: Observer?): NativeMst5.Observer? = observer?.let { original -> object : NativeMst5.Observer { override fun isCancelled() = original.isCancelled(); override fun onProgress(completed: Long, total: Long) = original.onProgress(completed, total) } }
}
